pkg/ctl/create/cluster.go

```diff
@@ -103,69 +103,40 @@ func createClusterCmd(g *cmdutils.Grouping) *cobra.Command {
 }
 
 func doCreateCluster(p *api.ProviderConfig, cfg *api.ClusterConfig, nameArg string, cmd *cobra.Command) error {
-	meta := cfg.Metadata
-
-	printer := printers.NewJSONPrinter()
-
-	if err := api.Register(); err != nil {
-		return err
-	}
 
 	ngFilter := NewNodeGroupFilter()
 
-	if clusterConfigFile != "" {
-		if err := eks.LoadConfigFromFile(clusterConfigFile, cfg); err != nil {
-			return err
-		}
-		meta = cfg.Metadata
-
-		incompatibleFlags := []string{
-			"name",
-			"tags",
-			"zones",
-			"version",
-			"region",
-			"nodes",
-			"nodes-min",
-			"nodes-max",
-			"node-type",
-			"node-volume-size",
-			"node-volume-type",
-			"max-pods-per-node",
-			"node-ami",
-			"node-ami-family",
-			"ssh-access",
-			"ssh-public-key",
-			"node-private-networking",
-			"node-security-groups",
-			"node-labels",
-			"node-zones",
-			"asg-access",
-			"external-dns-access",
-			"full-ecr-access",
-			"storage-class",
-			"vpc-private-subnets",
-			"vpc-public-subnets",
-			"vpc-cidr",
-			"vpc-from-kops-cluster",
-		}
-
-		for _, f := range incompatibleFlags {
-			if cmd.Flag(f).Changed {
-				return fmt.Errorf("cannot use --%s when --config-file/-f is set", f)
-			}
-		}
-
-		if meta.Name == "" {
-			return fmt.Errorf("metadata.name must be set")
-		}
-
-		if meta.Region == "" {
-			return fmt.Errorf("metadata.region must be set")
-		}
-
-		p.Region = meta.Region
-
+	cfgLoader := cmdutils.NewClusterConfigLoader(p, cfg, clusterConfigFile, cmd)
+
+	cfgLoader.FlagsIncompatibleWithConfigFile.Insert(
+		"tags",
+		"zones",
+		"nodes",
+		"nodes-min",
+		"nodes-max",
+		"node-type",
+		"node-volume-size",
+		"node-volume-type",
+		"max-pods-per-node",
+		"node-ami",
+		"node-ami-family",
+		"ssh-access",
+		"ssh-public-key",
+		"node-private-networking",
+		"node-security-groups",
+		"node-labels",
+		"node-zones",
+		"asg-access",
+		"external-dns-access",
+		"full-ecr-access",
+		"storage-class",
+		"vpc-private-subnets",
+		"vpc-public-subnets",
+		"vpc-cidr",
+		"vpc-from-kops-cluster",
+	)
+
+	cfgLoader.ValidateWithConfigFile = func() error {
```

Comment:
I think I would move this validation functionality into the ClusterConfigLoader. It should be the config loader the one who knows how to validate stuff, IMO, not the caller. I should be able to delegate everything related to configuration loader to an "entity" that does it for me. Otherwise it looks like a leaky abstraction to me. What do you think?

Reply:
I agree, however currently the validation logic differers slightly in each of the commands, hence it was does as a custom callback. I agree that it's leaky at present.
I've tried to do something like `cfgLoader.ValidateWithConfigFile = cmdutils.ValidateCreateClusterWithConfigFile`, but wasn't sure what exactly was the benefit of that.
I suppose mapping commands to validator implementations inside of `cmdutils.ClusterConfigLoader` maybe an okay thing to do and we could move the lists of flags there also, as the bulk of the actually flags is already defined in `cmdutils` package.

Reply:
We could do something like
```
var validatorsWithConfigFile = map[string]func() error{
	"eksctl create cluster": func() error {
		// do stuff
		return nil
	},
}

// NewClusterConfigLoader constructs a standard loader
func NewClusterConfigLoader(provider *api.ProviderConfig, spec *api.ClusterConfig, clusterConfigFile string, cmd *cobra.Command) *ClusterConfigLoader {
	nilValidatorFunc := func() error { return nil }

	logger.Debug("cmd.CommandPath = %s", cmd.CommandPath())
	l := &ClusterConfigLoader{
		provider: provider,
		path:     clusterConfigFile,
		cmd:      cmd,

		Spec: spec,

		ValidateWithConfigFile:             nilValidatorFunc,
		FlagsIncompatibleWithConfigFile:    sets.NewString("name", "region", "version"),
		ValidateWithoutConfigFile:          nilValidatorFunc,
		FlagsIncompatibleWithoutConfigFile: sets.NewString(),
	}

	cmdPath := cmd.CommandPath()

	if validatorFunc, ok := validatorsWithConfigFile[cmdPath]; ok {
		l.ValidateWithConfigFile = validatorFunc
	}

	return l
}
```

Reply:
I think this would be the classic example of one interface with multiple implementations. So we could have one interface ConfigLoader, and then CreateClusterConfigLoader whose validation methods are specific for that. The rest of the common code, `Load()` and `LoadMetadata` would be delegated to the common config loader.

Reply:
Yeah, definitely a good idea! Will do that, thanks :)